tests/v2/validation/upgrade/config.go

```diff
@@ -0,0 +1,156 @@
+//     for choosing some clusters to upgrade (version is optional, by default, empty string skips the test
+//     and "latest" upgrades to the latest available.)
+func loadUpgradeKubernetesConfig(client *rancher.Client) (clusters []Clusters, err error) {
+	upgradeConfig := new(Config)
+	config.LoadConfig(ConfigurationFileKey, upgradeConfig)
+
+	isConfigEmpty := len(upgradeConfig.Clusters) == 0
+	isFlagDeclared := client.Flags.GetValue(environmentflag.KubernetesUpgradeAllClusters)
+
+	if isConfigEmpty && !isFlagDeclared {
+		return clusters, errors.Wrap(err, "config doesn't match the requirements")
+	}
+
+	isUpgradeAllClusters := isFlagDeclared && isConfigEmpty
+	isUpgradeSomeClusters := !isConfigEmpty && !isUpgradeAllClusters
+
+	if isUpgradeAllClusters {
+		clusterList, err := listDownstreamClusters(client)
+		if err != nil {
+			return clusters, errors.Wrap(err, "couldn't list clusters")
+		}
+
+		for i := range clusterList {
+			cluster := new(Clusters)
+
+			cluster.Name = clusterList[i]
+			cluster.isLatestVersion = true
+
+			clusters = append(clusters, *cluster)
+		}
+	} else if isUpgradeSomeClusters {
+		for _, c := range upgradeConfig.Clusters {
+			cluster := new(Clusters)
+
+			cluster.Name = c.Name
+			cluster.VersionToUpgrade = c.VersionToUpgrade
+
+			isVersionFieldLatest := cluster.VersionToUpgrade == latestKey
+			if isVersionFieldLatest {
+				cluster.isLatestVersion = true
+			}
+
+			isUpgradeDisabled := cluster.VersionToUpgrade == ""
+			if isUpgradeDisabled {
+				cluster.isUpgradeDisabled = true
+			}
+
+			clusters = append(clusters, *cluster)
+		}
+	}
+
+	return
+}
+
+func loadUpgradeWorkloadConfig(client *rancher.Client) (clusters []Clusters, err error) {
+	upgradeConfig := new(Config)
+	config.LoadConfig(ConfigurationFileKey, upgradeConfig)
+
+	isConfigEmpty := len(upgradeConfig.Clusters) == 0
+	isFlagDeclared := client.Flags.GetValue(environmentflag.WorkloadUpgradeAllClusters)
+
+	if isConfigEmpty && !isFlagDeclared {
+		return clusters, errors.Wrap(err, "config doesn't match the requirements")
+	}
+
+	isUpgradeAllClusters := isFlagDeclared && isConfigEmpty
+	isUpgradeSomeClusters := !isConfigEmpty && !isUpgradeAllClusters
+
+	if isUpgradeAllClusters {
+		clusterList, err := listDownstreamClusters(client)
+		if err != nil {
+			return clusters, errors.Wrap(err, "couldn't list clusters")
+		}
+
+		for i := range clusterList {
+			cluster := new(Clusters)
+
+			cluster.Name = clusterList[i]
+			ingress := false
+			chart := false
+			cluster.FeaturesToTest = Features{
+				Ingress: &ingress,
+				Chart:   &chart,
+			}
+
+			clusters = append(clusters, *cluster)
+		}
+	} else if isUpgradeSomeClusters {
+		for _, c := range upgradeConfig.Clusters {
+			cluster := new(Clusters)
+
+			cluster.Name = c.Name
+			cluster.FeaturesToTest = c.FeaturesToTest
+
+			clusters = append(clusters, *cluster)
+		}
+	}
+
+	return
+}
+
+func listDownstreamClusters(client *rancher.Client) (clusterNames []string, err error) {
```

Comment:
Same question, is listDownstreamClusters potentially reusable in other tests?

Reply:
I don't have a strong opinion on this, but I feel like it might be better as private. 

We usually don't want to limit/encapsulate API calls and export as an extension if it's a single call. This PR contains some removed snippets specifically for the deamonsets and deployments creations, they're been removed because of the same reason. 

But I can also see this might create a code duplication, therefore it totally depends on the future usage. Not a rule or something, but we tend to create public extensions if they're used thrice